python/rmm/_lib/device_buffer.pyx

```diff
@@ -323,21 +323,21 @@ cpdef DeviceBuffer to_device(const unsigned char[::1] b, uintptr_t stream=0):
 @cython.boundscheck(False)
 cpdef void copy_ptr_to_host(uintptr_t db,
                             unsigned char[::1] hb,
-                            uintptr_t stream=0) nogil except *:
```

Comment:
here as well

Reply:
Because you can't compare to None without the gil

Reply:
Got it. I missed that you're explicitly releasing the GIL while synchronizing so we should be fine still.

Reply:
If we defaulted this to a singleton of the default `CudaStreamView` constructor. Could we easily checked if the stream is the legacy default stream versus per thread default stream and synchronize accordingly only when it's the legacy default stream?

Reply:
@jakirkham is this or the other functions used anywhere in ucx-py that we'd be concerned by this wanting to acquire the gil to call it?

Reply:
I'm not familiar with how to do "singleton of the default CudaStreamView constructor" in Python/Cython. Enlighten me? :)

Reply:
I tried making this take a default-constructed CudaStreamView but that also requires keeping the gil AFAICT.

Reply:
Sorry! Python initializes default arguments only once, so if we had `CudaStreamView stream=CudaStreamView()` it would only run that once and reuse the instance on each invocation.

That being said, we've taken the stance that the libraries should be able to be imported without a GPU and without initializing the CUDA driver / a CUDA context, so this would be problematic.

Reply:
Why would it be problematic? cuda_stream_view is a lightweight non-owning view. It's constructor does not call any CUDA APIs.

Reply:
This is what I'm trying now. The only question is moving the `nogil` inside the functions rather than making it part of the interface.

Reply:
> @jakirkham is this or the other functions used anywhere in ucx-py that we'd be concerned by this wanting to acquire the gil to call it?

I think this is used by cuDF (at the Python level) more directly that UCX-Py or Dask-CUDA at this stage (the latter used it before IIRC). After adding `"dask"` serialization Dask-CUDA relies on that to move data to host, which is what cuDF implements.

Reply:
Committed some improvements. Looks much cleaner to me now.

Reply:
@jakirkham how much overhead would there be from my changes? All it does is make the function interface require the GIL (because it calls the `CudaStreamView` constructor for the default parameter value, which requires the GIL), but inside the function everything is inside `with nogil`.